### ISSUE TITLE
s3: Unnecessary bucket_check condition while getting an object from s3

### ISSUE THREAD

**Issue by gmpuran** [ansible/ansible-modules-core#69]
Hi There,

I was advised here: https://github.com/ansible/ansible/issues/8123 to open a separate issue on newly separated repos.

Summary:

AWS IAM policy enables us to restrict users to specific file or folder in a bucket. While S3 get object operation is successful from other tools, ansible s3 module fails and returns an error Target bucket not found.

The operation succeeds when list bucket rights are provided to that IAM user.

Thanks!

**Comment by Ladret** [comment]
To work around this issue you'll need to add a bucket policy to allow the IAM user to list the bucket.

This is in addition to the policy to allow get acccess on the contents of the bucket

Example:

```
{
            "Sid": "Stmt1412808555554",
            "Action": [
                "s3:ListBucket"
            ],
            "Effect": "Allow",
            "Resource": "arn:aws:s3:::some-bucket",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::1234567890123:user/ansible-iam"
                ]
            }
        }
```

**Comment by gmpuran** [comment]
Thanks @Ladret , yes I'm aware of it.

By "The operation succeeds when list bucket rights are provided to that IAM user.", I meant the same.

**Comment by juliedavila** [comment]
@gmpuran could you do me a favor and format this issue like seen here: https://raw.githubusercontent.com/ansible/ansible/devel/ISSUE_TEMPLATE.md

I think I understand your issue but want to be 100% clear before working on a patch

**Comment by gmpuran** [comment]
Here you go. As the issue quite old, haven't got a chance to grab version details, but have provided detailed description of the issue.

##### Issue Type:

Bug Report

##### Ansible Version:

##### Environment:

N/A

##### Summary:

While getting an s3 object using ansible s3 module, explicit S3:ListBucket permission is required in IAM, for GET operation to be successful. This additional listing permission should not be required as one can restrict a particular IAM user to specific file or folder in S3 bucket.

##### Steps To Reproduce:
1. Create S3 bucket - ansible-s3
2. Create folder test/ and a file inside that folder test.txt
3. GET operation command will be:
   
   #### Simple GET operation
   - s3: bucket=ansible-s3 object=/test/test.txt dest=/usr/local/test.txt mode=get
4. If IAM policy is kept as follows:

``` json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1426510138000",
            "Effect": "Allow",
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws:s3:::ansible-s3/test"
            ]
        },
        {
            "Sid": "Stmt1426510160000",
            "Effect": "Allow",
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws:s3:::ansible-s3/test/*"
            ]
        }
    ]
}
```
1. Above GET operation will fail, unless S3:ListBucket operation access is provided to this IAM user.

##### Expected Results:

The GET operation should succeed without requiring S3:ListBucket rights.

##### Actual Results:

The GET operation fails unless IAM user has S3:ListBucket rights.

**Comment by owidjaya** [comment]
This problem seems to be related to boto. update to latest boto. add endpoints_path in boto config file to override the endpoints that is problematic. you can find the endpoints that amazon define in here http://docs.aws.amazon.com/general/latest/gr/rande.html. 
here is the original endpoints.json file for boto that contain all the endpoints definition https://github.com/boto/boto/blob/develop/boto/endpoints.json.

**Comment by magnetik** [comment]
To me this bug shouln't be closed: the `s3:ListBucket` is still required.

**Comment by bcoca** [comment]
this bug has not been closed, that was another ticket

**Comment by jonathanhle** [comment]
+1 just ran into this.

**Comment by juziel** [comment]
I just ran into this issue after updating from 1.9.1 to 1.9.4. The role wasn't changed, but the issue suddenly started occurring.

**Comment by JoelGilliland** [comment]
@juziel I doubt this was caused by updating, more likely the issue lies in a change made to permissions pertaining to AWS user/bucket you're trying to access. The work around is to allow the list permissions to S3, not ideal.

**Comment by juliedavila** [comment]
Sorry it has taken very long to address this, looking into it now

**Comment by juliedavila** [comment]
Pull request with a fix is here: https://github.com/ansible/ansible-modules-core/pull/3248

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/30234
